helix-core/src/main/java/org/apache/helix/controller/stages/PerReplicaThrottleStage.java

```diff
@@ -270,4 +589,23 @@ public void setPriority(String priority) {
       }
     }
   }
+
+  private boolean isDownwardTransition(StateModelDefinition stateModelDef, Message message) {
```

Comment:
I feel you can use some variance of this method to calculate the propagate state count map. If this is the case, we can move this method to the state model definition (or a util class) to cleanup the code here. More importantly, the class will provide better leverage.

Reply:
Let me make sure I understand.

1/ " feel you can use some variance of this method to calculate the propagate state count map"? This is to test if a message is downward. Seems to me, totally irrelevant to calculate the propagate state count map. Can you elaborate? Maybe some code. I totally don't get it.

2/ For sure I can move this one to stateModel definition. 

Let me sync with your over slack.

Reply:
moved to StateModelDefinition.

Reply:
StateModelDefinition.isDownwardTransition(message), just like that.

Reply:
There is a logic in the PR to find all the related state for propagation, right? I was just thinking if we can let these 2 methods share some common components to reduce duplication. To be more specific, we can add a method to StateModelDefinition which return all the down stream or upper stream states given a certain state. Then the propagation count calculation can use this method to find all the propagate target, and this downward transition method can use the same method to find if the to state is one of the down stream state of the from state.